### PR TITLE
442/bugfix/api status queries trigger multiple report generations

### LINKED ISSUE #442: `/api` status queries can trigger multiple report generation
**Describe the bug**
Probably relates to #441.

When accessing the Island, the `api` query checks if a report has been generated and can cause it to get generated - and if that's slow it can get stuck in a request loop and bottleneck the HTTP server. 

**To Reproduce**
Same as #441.

**Expected behavior**
`/api` shouldn't cause any activity on the server - command/query separation is a must.

### PR DESCRIPTION
**Pull request by ShayNehmad**
# Feature / Fixes
Fixes #442 .

* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Have you successfully tested your changes locally?

## Log transcript of the feature working
With five seconds of sleep in report generation. *Logs have been deleted from the code*. 
```
2019-10-02 13:20:20,802 - root.py:65 - get_completed_steps() - DEBUG - Infection is done; checking if should generate report...
2019-10-02 13:20:20,803 - root.py:81 - generate_report() - DEBUG - Generating report from Root service.
SLEEP
DONE SLEEP
2019-10-02 13:20:25,805 - report.py:558 - get_domain_issues() - INFO - Domain issues generated for reporting
2019-10-02 13:20:25,894 - report.py:237 - get_azure_creds() - INFO - Azure machines creds generated for reporting
2019-10-02 13:20:25,894 - report.py:111 - get_azure_issues() - INFO - Azure issues generated for reporting
2019-10-02 13:20:25,895 - report.py:593 - get_issues() - INFO - Issues generated for reporting
2019-10-02 13:20:25,901 - config.py:164 - get_default_config() - INFO - Default config was called
2019-10-02 13:20:25,999 - report.py:144 - get_scanned() - INFO - Scanned nodes generated for reporting
2019-10-02 13:20:26,000 - report.py:168 - get_exploited() - INFO - Exploited nodes generated for reporting
2019-10-02 13:20:26,000 - report.py:194 - get_stolen_creds() - INFO - Stolen creds generated for reporting
2019-10-02 13:20:26,002 - report.py:237 - get_azure_creds() - INFO - Azure machines creds generated for reporting
2019-10-02 13:20:26,055 - attack_report.py:67 - generate_new_report() - ERROR - Attack technique does not have it's report component added to attack report service. u'T1078'
2019-10-02 13:20:26,101 - root.py:87 - generate_report() - DEBUG - Finished generating report from Root service.
2019-10-02 13:20:26,102 - wsgi.py:374 -       _log() - INFO - 200 GET /api (127.0.0.1) 5406.00ms
2019-10-02 13:20:26,457 - wsgi.py:374 -       _log() - INFO - 200 GET /api/netmap (127.0.0.1) 345.00ms
2019-10-02 13:20:26,467 - wsgi.py:374 -       _log() - INFO - 200 GET /api/telemetry-feed?timestamp=null (127.0.0.1) 9.00ms
2019-10-02 13:20:26,566 - root.py:65 - get_completed_steps() - DEBUG - Infection is done; checking if should generate report...
2019-10-02 13:20:26,571 - wsgi.py:374 -       _log() - INFO - 200 GET /api (127.0.0.1) 98.00ms
2019-10-02 13:20:27,174 - wsgi.py:374 -       _log() - INFO - 200 GET /api/netmap (127.0.0.1) 356.00ms
2019-10-02 13:20:27,272 - root.py:65 - get_completed_steps() - DEBUG - Infection is done; checking if should generate report...
2019-10-02 13:20:27,275 - wsgi.py:374 -       _log() - INFO - 200 GET /api (127.0.0.1) 96.00ms
2019-10-02 13:20:27,721 - root.py:65 - get_completed_steps() - DEBUG - Infection is done; checking if should generate report...
2019-10-02 13:20:27,724 - wsgi.py:374 -       _log() - INFO - 200 GET /api (127.0.0.1) 91.00ms
2019-10-02 13:20:28,174 - root.py:65 - get_completed_steps() - DEBUG - Infection is done; checking if should generate report...
2019-10-02 13:20:28,177 - wsgi.py:374 -       _log() - INFO - 200 GET /api (127.0.0.1) 95.00ms
2019-10-02 13:20:30,795 - root.py:65 - get_completed_steps() - DEBUG - Infection is done; checking if should generate report...
```

##  Changes
  - Added event to Root service to make sure report isn't generated unnecessarily, and to support future multithreaded server. 
  - Improved the logic of the Root Service `get_server_info()` function.

### CODE REVIEW

**Comment by danielguardicore** [comment]
I don't understand how the logs show what you want it to show. get_completed_steps is only called again after report generation was finished.

File: monkey/infection_monkey/network/network_scanner.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 from infection_monkey.model.victim_host_generator import VictimHostGenerator
 from infection_monkey.network.info import local_ips, get_interfaces_ranges
 from infection_monkey.network import TcpScanner, PingScanner
-from infection_monkey.utils import is_windows_os
```
Comment (danielguardicore) [inline monkey/infection_monkey/network/network_scanner.py]: Why did this change in this PR?

**Comment by danielguardicore** [comment]
Not good enough. People can also query the report and attackreport directly. The lock needs to be moved to protect the actual report generation themselves.

File: monkey/infection_monkey/network/network_scanner.py (hunk 1 above)
Comment (ShayNehmad) [reply monkey/infection_monkey/network/network_scanner.py]: Should have been a different branch, probably committed by mistake. Removed

**Comment by ShayNehmad** [comment]
When querying the report directly via `ReportService`, it checks if the report exists: 
https://github.com/guardicore/monkey/blob/32e98fa41879c277552465f119ecc21a84f506bc/monkey/monkey_island/cc/services/reporting/report.py#L780

That way we're protected if it was created beforehand and if someone asks for the report they only get it once - but the "freshest" they can. 

Are you suggesting we protect that API (`api/report/security`) as well, and with the same lock? @danielguardicore

File: monkey/monkey_island/cc/services/reporting/report_generation_synchronisation.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import logging
+import threading
+
+logger = logging.getLogger(__name__)
+
+# These are pseudo-singletons - global Locks. These locks will allow only one thread to generate a report at a time.
+# Report generation can be quite slow if there is a lot of data, and the UI queries the Root service often; without
+# the locks, these requests would accumulate, overload the server, eventually causing it to crash.
+logger.debug("Initializing report generation locks.")
+__report_generating_lock = threading.Semaphore()
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/reporting/report_generation_synchronisation.py]: Why do we need this third lock? Just for readability?

**Pull request merged**